sigv4-signing/go/Sigv4Signer.go

```diff
@@ -0,0 +1,239 @@
+
+	// Test Credentials for sample connection verification. SessionToken can be empty string if using non-temporary credentials.
+	accessKey := "YourAccessKey"
+	secretKey := "YourSecretKey"
+	sessionToken := "YourSessionToken"
+
+	// AWS Region (e.g. us-west-2)
+	region := "YourChannelRegion"
+
+	////////////////////////////////////////////
+
+
+	// Compute SigV4 signing
+	// Ref: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+	// https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-android/blob/master/src/main/java/com/amazonaws/kinesisvideo/utils/AwsV4Signer.java
+
+	signedUri := AwsV4Signer{}.sign(uri, accessKey, secretKey, sessionToken, wssUri, region)
+	println("Go SignedURL log for verifications (e.g. using wscat ) =", signedUri.String())
+
+	// Connect to the SigV4 signed Endpoint
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	c, resp, err := websocket.DefaultDialer.Dial(signedUri.String(), nil)
+
+	// Print the responses from the connection
+	println("Response code=", resp.StatusCode)
+	if err != nil {
+		log.Fatalln("Error in connection: ", err)
+	}
+
+	b, errBodyResp := io.ReadAll(resp.Body)
+	if errBodyResp != nil {
+		log.Fatalln("Error in reading response body: ", errBodyResp)
+	}
+
+	fmt.Println("Response body", string(b))
+
+	body, errResp := httputil.DumpResponse(resp, true)
+	if errResp != nil {
+		log.Fatal("Error printing response body:", err)
+	}
+
+	fmt.Println(string(body))
+
+	defer c.Close()
+
+}
+
+func (signer AwsV4Signer) sign(uri *url.URL, accessKey, secretKey, sessionToken string, wssURI *url.URL, region string) *url.URL {
+
+	dateMilli := time.Now().UnixMilli()
+	amzDate := getTimeStamp(dateMilli)
+	datestamp := getDateStamp(dateMilli)
+
+	queryParamsMap := buildQueryParamsMap(uri, accessKey, sessionToken, region, amzDate, datestamp)
+	canonicalQuerystring := getCanonicalizedQueryString(queryParamsMap)
+	canonicalRequest := getCanonicalRequest(uri, canonicalQuerystring)
+	stringToSign := signString(amzDate, createCredentialScope(region, datestamp), canonicalRequest)
+
+	signatureKey := getSignatureKey(secretKey, datestamp, region, SERVICE)
+	signature := hex.EncodeToString(hmacSha256(stringToSign, signatureKey))
+
+	signedCanonicalQueryString := canonicalQuerystring + "&" + X_AMZ_SIGNATURE + "=" + signature
+
+	signedURI, err := url.Parse(wssURI.Scheme + "://" + wssURI.Host + getCanonicalURI(uri) + "?" + signedCanonicalQueryString)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	return signedURI
+}
+
+func buildQueryParamsMap(uri *url.URL, accessKey, sessionToken, region, amzDate, datestamp string) map[string]string {
+
+	xAmzCredential := urlEncode(accessKey + "/" + createCredentialScope(region, datestamp))
+
+	queryParamsMap := map[string]string{
+		X_AMZ_ALGORITHM:  ALGORITHM_AWS4_HMAC_SHA_256,
+		X_AMZ_CREDENTIAL: xAmzCredential,
+		X_AMZ_DATE:       amzDate,
+		X_AMZ_EXPIRES:    "299",
```

Comment:
// The SigV4 signer has a  maximum time limit of five minutes; Once a connection is established, Peers exchange signaling messages, and the P2P connection is successful, the media P2P session can continue for longer period of time.

Reply:
Done. added to the go sample as well.